### ISSUE TITLE
When used/requied Locale::*** modules unexpectedly modify $_

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5928]
Migrated from [rt.perl.org#17340](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=17340) (status was 'resolved')

Searchable as RT17340$

**Comment by p5pRT** [comment]
### From @IlyaM

#### Created by @IlyaM



When used/requied Locale&#8203;::\*\*\* modules unexpectedly modify $\_\. It can
cause problems as described in

&nbsp;    http&#8203;://www\.perlmonks\.com/index\.pl?node\_id=195926

Following patch \(against 5\.8\.0\) fixes this problem&#8203;:

<details><summary>Inline Patch</summary>

```diff
diff -u -r perl-5.8.0-orig/lib/Locale/Country.pm perl-5.8.0/lib/Locale/Country.pm
--- perl-5.8.0-orig/lib/Locale/Country.pm	Sun Sep 15 22:52:48 2002
+++ perl-5.8.0/lib/Locale/Country.pm	Sun Sep 15 22:54:43 2002
@@ -267,6 +267,7 @@
     my ($alpha2, $alpha3, $numeric);
     my ($country, @countries);
 
+    local $_;
 
     while (<DATA>)
     {
diff -u -r perl-5.8.0-orig/lib/Locale/Currency.pm perl-5.8.0/lib/Locale/Currency.pm
--- perl-5.8.0-orig/lib/Locale/Currency.pm	Sun Sep 15 22:52:49 2002
+++ perl-5.8.0/lib/Locale/Currency.pm	Sun Sep 15 22:55:13 2002
@@ -108,6 +108,7 @@
     my $code;
     my $currency;
 
+    local $_;
 
     while (<DATA>)
     {
diff -u -r perl-5.8.0-orig/lib/Locale/Language.pm perl-5.8.0/lib/Locale/Language.pm
--- perl-5.8.0-orig/lib/Locale/Language.pm	Sun Sep 15 22:52:49 2002
+++ perl-5.8.0/lib/Locale/Language.pm	Sun Sep 15 22:55:00 2002
@@ -107,6 +107,7 @@
     my $code;
     my $language;
 
+    local $_;
 
     while (<DATA>)
     {
diff -u -r perl-5.8.0-orig/lib/Locale/Script.pm perl-5.8.0/lib/Locale/Script.pm
--- perl-5.8.0-orig/lib/Locale/Script.pm	Sun Sep 15 22:52:49 2002
+++ perl-5.8.0/lib/Locale/Script.pm	Sun Sep 15 22:55:26 2002
@@ -160,6 +160,7 @@
     my ($alpha2, $alpha3, $numeric);
     my $script;
 
+    local $_;
 
     while (<DATA>)
     {
```
</details>


<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.8.0:

Configured by ilya at Sat Jul 20 10:28:51 MSD 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.16, archname=i686-linux-thread-multi
    uname='linux abra.ru 2.4.16 #1 ÐÎÄ ÄÅË 10 20:32:46 msk 2001 i686 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.4 20011006 (Debian prerelease)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.2.4.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.2.4'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /home/ilya/perl-lib
    /home/ilya/perl580/lib/5.8.0/i686-linux-thread-multi
    /home/ilya/perl580/lib/5.8.0
    /home/ilya/perl580/lib/site_perl/5.8.0/i686-linux-thread-multi
    /home/ilya/perl580/lib/site_perl/5.8.0
    /home/ilya/perl580/lib/site_perl
    .


Environment for perl v5.8.0:
    HOME=/home/ilya
    LANG=ru_RU.KOI8-R
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/ilya/bin:/home/ilya/Office51/bin:/usr/local/bin:/usr/bin:/bin:/usr/X11R6/bin:/usr/games:/home/ilya/arch/bin
    PERL5LIB=/home/ilya/perl-lib
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

\(Ilya Martynov\) \(via RT\) wrote&#8203;:
> When used/requied Locale&#8203;::\*\*\* modules unexpectedly modify $\_\. It can
> cause problems as described in
> 
>     http&#8203;://www\.perlmonks\.com/index\.pl?node\_id=195926
> 
> Following patch \(against 5\.8\.0\) fixes this problem&#8203;:

Thanks\, applied as \#17908\.

**Comment by p5pRT** [comment]
@rspier - Status changed from 'new' to 'resolved'
